sympy/matrices/determinant.py

```diff
@@ -410,10 +412,24 @@ def _charpoly(M, x='lambda', simplify=_simplify):
             m = m * (x - simplify(i))
         return PurePoly(m, x)
 
-    berk_vector = _berkowitz_vector(M)
-    x = uniquely_named_symbol(x, berk_vector, modify=lambda s: '_' + s)
+    t = Dummy('t')
+    if use_domain is False:
+        berk_vector = _berkowitz_vector(M)
+        dummy_poly = Poly(berk_vector, t)
+        x = uniquely_named_symbol(x, dummy_poly, modify=lambda s: '_' + s)
+        return PurePoly([simplify(a) for a in berk_vector], x)
+
+    DOM = DomainMatrix.from_Matrix(M)
+    domain = DOM.domain
+    dom_vector = _charpoly_DOM(DOM)
+    dummy_poly = PurePoly(dom_vector, t, domain=domain)
+    x = uniquely_named_symbol(x, dummy_poly, modify=lambda s: '_' + s)
+
+    p = PurePoly(dom_vector, x, domain=domain)
+    if domain.is_FractionField and not all(t.is_Symbol for t in domain.symbols):
+        return PurePoly(simplify(p.as_expr()), x)
```

Comment:
This is messy and repetitive. This code can be better organised.

Reply:
Indeed, the code looks messy and repetitive.

Maybe we can define an `function` that can give us the `uniquely_named_symbol`.
Something like:
```py
def get_uniquely_named_symbol(vector, x, domain):
    t = Dummy('t')
    if domain is None:
        dummy_poly = Poly(vector, t)
    else:
        dummy_poly = PurePoly(vector, t, domain=domain)
    
    x = uniquely_named_symbol(x, dummy_poly, modify=lambda s: '_' + s)
    return x
```
This can be used by both the methods of `charpoly` i.e. using domain and not.

@oscarbenjamin, Kindly share your opinion.
Thanks.

Reply:
Maybe it can't be improved that much. I suggest something like this:
```python
    xdum = Dummy('t')

    if use_domain:
        DOM = DomainMatrix.from_Matrix(M)
        domain = DOM.domain
        dom_vector = _charpoly_DOM(DOM)
        dummy_poly = PurePoly(dom_vector, xdum, domain=domain)

        # XXX: calling simplify like this is a bad idea.
        # The caller can simplify if they want
        if any(t.is_Symbol for t in getattr(domain, 'symbols', ())):
            dummy_poly = PurePoly(simplify(dummy_poly.as_expr()), xdum)
    else:
        berk_vector = _berkowitz_vector(M)
        berk_vector = [simplify(a) for a in berk_vector]
        dummy_poly = PurePoly(berk_vector, xdum)

    xsym = uniquely_named_symbol(x, dummy_poly, modify=lambda s: '_' + s)
    poly = dummy_poly.subs(xdum, xsym)
```

Reply:
Sure, this looks more readable and well structured.
Thanks.